adam-core/src/main/scala/org/bdgenomics/adam/rdd/GenomicRDD.scala

```diff
@@ -539,6 +639,70 @@ trait GenomicRDD[T, U <: GenomicRDD[T, U]] {
   }
 
   /**
+   * Prepares two RDDs to be joined with any shuffleRegionJoin. This includes copartition
+   * and sort of the rightRdd if necessary.
+   *
+   *  @param genomicRdd The RDD to join to.
+   *  @return A tuple containing both RDDs ready to join, the partitionMap and the number
+   *          of partitions.
+   *          _._1 is the leftRdd,
+   *          _._2 is the rightRdd,
+   *          _._3 is the partitionMap for the joined RDDs,
+   *          _._4 is the number of partitions
+   */
+  private def prepareForShuffleRegionJoin[X, Y <: GenomicRDD[X, Y], Z <: GenomicRDD[(T, X), Z]](
+    genomicRdd: GenomicRDD[X, Y],
+    optPartitions: Option[Int] = None)(
+      implicit tTag: ClassTag[T], xTag: ClassTag[X]): (RDD[(ReferenceRegion, T)], RDD[(ReferenceRegion, X)], Option[Seq[Option[(ReferenceRegion, ReferenceRegion)]]], Int) = {
+
+    // did the user provide a set partition count?
+    // if no, take the max partition count from our rdds
+    val partitions = optPartitions.getOrElse(this.rdd.partitions.length)
+
+    val leftRdd: GenomicRDD[T, U] = if (sorted) {
+      if (partitions != rdd.partitions.length) {
+        evenlyRepartition(partitions)
+      } else {
+        this
+      }
+    } else {
+      repartitionAndSort(partitions)
+    }
+
+    assert(leftRdd.sorted)
+    val rightRdd: GenomicRDD[X, Y] = genomicRdd.copartitionByReferenceRegion(leftRdd)
+    assert(rightRdd.sorted)
+
+    // Partition Map for the left side
+    val leftPartitionMap = leftRdd.partitionMap.get
+    val leftRddToJoin = leftRdd.rdd.mapPartitionsWithIndex((idx, iter) => {
+      leftRdd.extractReferenceRegionsForPartition(iter, leftPartitionMap(idx)).map(f => (f._1, f._2))
+    })
+
+    // Partition map for the right side
+    val rightPartitionMap = rightRdd.partitionMap.get
+
+    val rightRddToJoin = rightRdd.rdd.mapPartitionsWithIndex((idx, iter) => {
+      rightRdd.extractReferenceRegionsForPartition(iter, rightPartitionMap(idx)).map(f => (f._1, f._2))
+    })
+    assert(leftRddToJoin.partitions.length == rightRddToJoin.partitions.length,
+      "Partitions are not equal size!")
+
+    val partitionBoundsForJoin = Some(
```

Comment:
This is the new partition region bound data, right? Can we add an inline comment to document?